src/posix/platform/spi_interface.cpp

```diff
@@ -0,0 +1,833 @@
+    {
+        // We guard this with the above check because we don't want to overwrite any previously received frames.
+        PushPullSpi();
+    }
+}
+
+otError SpiInterface::WaitForFrame(struct timeval &aTimeout)
+{
+    otError  error   = OT_ERROR_NONE;
+    uint32_t timeout = kMsecPerDay;
+    fd_set   readFdSet;
+    int      ret;
+
+    FD_ZERO(&readFdSet);
+
+    if (mIntGpioValueFd >= 0)
+    {
+        if (CheckInterrupt())
+        {
+            // Interrupt pin is asserted, set the timeout to be 0.
+            timeout = 0;
+            otLogDebgPlat("Interrupt.");
+        }
+        else
+        {
+            // The interrupt pin was not asserted, so we wait for the interrupt pin to be asserted by adding it to the
+            // read set.
+            FD_SET(mIntGpioValueFd, &readFdSet);
+        }
+    }
+    else if (timeout > kSpiPollPeriodMs)
+    {
+        // In this case we don't have an interrupt, so we revert to SPI polling.
+        timeout = kSpiPollPeriodMs;
+    }
+
+    if (static_cast<uint64_t>(timeout) * US_PER_MS <
+        static_cast<uint64_t>(aTimeout.tv_sec) * US_PER_S + aTimeout.tv_usec)
+    {
+        aTimeout.tv_sec  = static_cast<time_t>(timeout / kMsecPerSec);
+        aTimeout.tv_usec = static_cast<suseconds_t>((timeout % kMsecPerSec) * kUsecPerMsec);
+    }
+
+    ret = select(mIntGpioValueFd + 1, &readFdSet, NULL, NULL, &aTimeout);
+    if ((ret < 0) && (errno != EINTR))
+    {
+        DieNow(OT_EXIT_ERROR_ERRNO);
+    }
+
+    if (FD_ISSET(mIntGpioValueFd, &readFdSet))
+    {
+        struct gpioevent_data event;
+
+        otLogDebgPlat("Interrupt.");
+
+        // Read event data to clear interrupt.
+        VerifyOrDie(read(mIntGpioValueFd, &event, sizeof(event)) != -1, OT_EXIT_FAILURE);
+    }
+
+    // If we can receive a packet.
+    if (CheckInterrupt())
+    {
+        PushPullSpi();
+    }
+
+    return error;
+}
+
+otError SpiInterface::SendFrame(const uint8_t *aFrame, uint16_t aLength)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aLength < (kMaxFrameSize - kSpiFrameHeaderSize), error = OT_ERROR_NO_BUFS);
+    VerifyOrExit(!mSpiTxIsReady, error = OT_ERROR_BUSY);
+
+    memcpy(&mSpiTxFrameBuffer[kSpiFrameHeaderSize], aFrame, aLength);
+
+    mSpiTxIsReady     = true;
+    mSpiTxPayloadSize = aLength;
+
+    PushPullSpi();
+
+exit:
+    return error;
+}
+
+void SpiInterface::HandleReceivedFrame(Ncp::SpiFrame &aSpiFrame)
+{
+    const uint8_t *spinelFrame = aSpiFrame.GetData();
+
+    for (uint16_t i = 0; i < aSpiFrame.GetHeaderDataLen(); i++)
+    {
+        if (mRxFrameBuffer.WriteByte(spinelFrame[i]) != OT_ERROR_NONE)
+        {
+            mRxFrameBuffer.DiscardFrame();
+            otLogNotePlat("No enough memory buffers, drop packet");
+            ExitNow();
+        }
+    }
+
+    mCallbacks.HandleReceivedFrame();
+
+exit:
+    return;
+}
+
+void SpiInterface::LogError(const char *aString)
+{
+    fprintf(stderr, "%s: %s", aString, strerror(errno));
```

Comment:
Not necessary, for we have `-v` argument which will print error to std error.

Reply:
Updated